### PR TITLE
Implement Editor commands

### PR DESCRIPTION
**Pull request by punkkeks**
This is a little pull request of a large series. The aim is to implement all editorcommands, so they can be edited in the gui. Currently only a few are implemented and there are some quirks, but it works.
I'd like to have some feedback to the implementation-style and other stuff, this won't be the last commit (hopefully) and you are welcome to contribute (also to my branch if you want)

### CODE REVIEW

**Comment by Ghabry** [comment]
I'm currently doing a development break after the Player 0.5 release. Will take a look next  weekend, sorry.

**Comment by Ghabry** [comment]
Nice contribution. I'm currently doing a development break after the Player 0.5 release. Will take a look next  weekend, sorry.

Just some essay as a discussion point:
tl;dr: Don't invest too much time yet in parsing com.parameters or setting it, the API will very likely change when I update liblcf.

Why is a change necessary, because:

That is not supported in easyprg yet (because liblcf needs rework) but what I dream of for our event commands is to get rid of arbitrary limitations. E.g. There is no obvious reason why some arguments only accept constants, but not variables. We want to be a better RPG Maker, Enterbrain had nothing revolutionary since 2000, MV is basicly the same as RPG2k just with bigger sprites.

Variable Operations: Why only variable reference when setting a single value? Why does Variable Range not support indirection for start and finish. Operands: There could be more "complex" OPs like Sqrt or ^n. What you can access for Item, Hero, Sprite and Other is totally arbitrary, too, you can't access most database fields with this.
In general everything that accepts a constant should accept a variable, too.

Even more fancy: Picture Operation is another thing, the only reason for that "PicPoitner Patch" is because the event commands lack string functions and Variable-references for picture IDs.

**Comment by punkkeks** [comment]
Thanks for that, I won't parse the commands any further (but will store them as member variable, just to be prepared in the future ;) )
I will do some more work on the UIs, don't know if I keep the original layout or not (but I don't have a clue how it could be improved :/ ). Have a nice week

**Comment by Ghabry** [comment]
Well I just somehow don't want the editor to look like some blatant chinese copy :D. But most of the dialogs could be reused yes. Even when we someday extend them they would be a good base...

What would be cool maybe (assuming the user understands this) would be to combine these
"Specific: [        v]" and "Stored in: [        v]" into one combobox which supports constants and variables.
Though no idea how to handle this intuitively. Maybe prefix the numbering with V for variables. Or do you think keeping them seperated is better? One could go even more crazy here and support an arbitrary amount of Vs, so that VV5 will use the Variable pointed to by the Value in Variable 5. Too much C-programming ;)

File: CMakeLists.txt [hunk 1]
```diff
@@ -57,6 +57,17 @@ set(SRCS
...
+    src/commands/changemoney.cpp
+    src/commands/changeparty.cpp
+    src/commands/changeexperience.cpp
+    src/commands/facegraphics.cpp
+    src/commands/inputnumber.cpp
+    src/commands/messageoptions.cpp
+    src/commands/showchoices.cpp
+    src/commands/showmessage.cpp
+    src/commands/switchoperations.cpp
+    src/commands/variableoperations.cpp
```
Comment (Ghabry) [inline CMakeLists.txt]: Wrong idendation

File: src/tools/rpgmodel.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#ifndef RPGMODEL_H
+#define RPGMODEL_H
+
+#include <QObject>
+#include <QAbstractItemModel>
+#include <data.h>
+
+template <class DATA>
```
Comment (Ghabry) [inline src/tools/rpgmodel.h]: Can you explain how this class DATA and class MODEL works?

File: src/tools/rpgmodel.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    return QVariant();
+}
+
+
+
+
+#define VARSTRUCT(STRUCTNAME, VARNAME) \
+    namespace detail { struct STRUCTNAME { decltype(Data::VARNAME) operator()() const { return Data::VARNAME; } }; } \
+    using  STRUCTNAME##RpgModel = RpgModel<detail::STRUCTNAME>;
```
Comment (Ghabry) [inline src/tools/rpgmodel.h]: And this

**Comment by MarianoGnu** [comment]
I hope we can sopport visual scripting at some point, current interpreter would do okay with a node based scripting language

File: src/tools/rpgmodel.h (hunk 2 above)
Comment (punkkeks) [inline src/tools/rpgmodel.h]: I don't want to write the same code all over again just because the underlying data changes (that is what DATA is for, that is the underlying data to use (a vector of things :), could also be std::vector<string> or whatever, it must be at least accessable by operator[](int), as done in line 37 and has size() in line 14.
So, DATA is a class, that has a operator()() (could also be an lambda within a lambda, it must be "callable" twice (or instanciate it first and then call it, as seen in this example). In line 21 I get the type of the second call and use decltype to deduce the type. In line 12 I assign _data. (I hope this step is clear). 
The two typedefs in 17 and 18 are for use in the modelclass, just typedefs, nothing special :)

File: src/tools/rpgmodel.h (hunk 3 above)
Comment (punkkeks) [inline src/tools/rpgmodel.h]: Here I define where/what the underlying data is. STRUCTNAME will be replaced by the first macro parameter (mostly the class inside the vector). This struct has an operator()() (why is explained above) that will return a reference to the RPGdata (just a return). The return value will be autodeduced by decltype.
Next line just says, that there will be (e.g.) a ActorRpgModel. the ## is stringconcat in preprocessor, and what using does is just declares something as something :D In this case you don't have to write `RpgModel<detail::Actor>` if you want an Model with an Actor, but instead just use ActorRpgModel, far more simple and hides the nasty stuff :) Hope this helps

File: src/tools/rpgcombobox.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#include <QObject>
+#include <QComboBox>
+#include <QAbstractItemModel>
+#include <QCompleter>
+#include <functional>
+#include "rpgmodel.h"
+#include "../dialogedit.h"
+
+
+template <class MODEL>
```
Comment (punkkeks) [inline src/tools/rpgcombobox.h]: Here you can pass any class, that is dereivec from QAbstractItemModel, in this case I pass the RpgModels I declared before to use them as a model for the combobox to make it easier to populate all comboboxes with data.

**Comment by punkkeks** [comment]
I'm not very good in explanations (regarding to my girlfriend), but I hope you will understand what I want to say (elsewise I could explain it a little bit better in german if you want ;) ).
I want to improve the combobox a little bit further, but the general code will stay the same. I Hope I can make it this week. Work is quiet busy this week :/

File: src/tools/rpgmodel.h (hunk 2 above)
Comment (punkkeks) [inline src/tools/rpgmodel.h]: Yep, that's it ;)
I love template programming, but why I use it here, is because we can use it easily in the editor now (when you open qt-editor and open the `variableoperations.ui` in line 49, you see `VariableRpgComboBox`. That is so neat, to use it like that, instead of using a generic qcombobox and insert all the data in the cpp file. You also could achieve this with subclassing, but hmm... don't like subclassing that much... one could argue about that

Comment (Ghabry) [inline src/tools/rpgmodel.h]: If you love template programming i recommend taking a look at reader_struct.cpp and reader_struct.h in liblcf >.<
Yeah implementing a specific combobox that is prefilled makes sense, no code duplication. I would have implemented it the same way (probably as a subclass), did this already in another Qt project.

**Comment by Ghabry** [comment]
@MarianoGnu 
My only experience with graphical scripting was Kismet in Unreal and I hated this so much ;)

@punkkeks 
I think I understood now how it works.
Somehow I'm also thinking too much in OOP, that's the result when the university only teaches you stupid Java -.-, but with templates you can generalize even if the class has no common parent (as for .ID and .name)... cool

**Comment by MarianoGnu** [comment]
@Ghabry RM2000/3's event system is visual scripting too

**Comment by Ghabry** [comment]
For me visual scripting is "place boxes and connect with lines".

In my opinion RM2k script is just a super verbose scripting language. You could map this to a text representation easily, the editor just doesn't allow it.
That's actually one of the things I'm planning for "liblcf vNext", a text version of the "script" language.

**Comment by MarianoGnu** [comment]
Use python, python is love :P

**Comment by punkkeks** [comment]
Uhhh python bindings for rpgm2kx :D awesome, but guys, let's get back to topic please ;)

**Comment by Ghabry** [comment]
Sorry for the long review. Will be more focused again now...

Looks good to me and the solution is clever, without any duplication. Works in the GUI properly, too. Just fix the indent in CMakeLists and we are done ;)

File: CMakeLists.txt (hunk 1 above)
Comment (punkkeks) [inline CMakeLists.txt]: Ha! I see now, that I don't have a wrong indention (.. I do, yes, but no!), but instead the other lines are indented with tabs, instead of spaces. Should I indent mine with tabs, or should I retab the other lines?

Comment (Ghabry) [inline CMakeLists.txt]: The problem here is that the Player uses Tabs everywhere but the Editor spaces... Probably because Qt Creator uses Spaces per default... That consistency.
Well, change all to spaces.

**Comment by punkkeks** [comment]
Sorry that it took so long, i have a lot of stuff todo right now (we're introducing IPv6 in our company right now, I'm very excited :3 ) I'm not sure when I can do the next UIs, but feel free to implement more and more and more :D

**Comment by Ghabry** [comment]
Was worried that you left us :D
